docs/adr/0016-publishing-times-and-change-history.md

```diff
@@ -0,0 +1,214 @@
+  fully realised);
+- change history is the lone outlier as user visible data on GOV.UK that a
+  publishing application cannot fix in case of any issues;
+- the Publishing API change note system is incompatible with the ability to
+  change the time content is marked as first published (as the change history
+  will disagree with this value);
+- the Publishing API is ill suited to provide mechanisms to rectify any issues
+  with change notes as there is no API applications can use.
+
+## Status
+
+Accepted
+
+## Consequences
+
+Content Publisher will store all the information needed to present
+`first_published_at`, `public_updated_at` and change history to the Publishing
+API. In doing so Content Publisher will be considered the canonical source
+of this data. This data will be reflected by non user editable data of
+`first_published_at` on a Document model and `published_at` on an Edition model,
+which reflect the times Content Publisher first sent content to the Publishing
+API. In addition to the existing `change_note` field on a MetadataRevision
+there will also be a `change_history` field that stores a JSON representation
+of all the previous change history for the content.
+
+This `change_history` field will be used to store a collection of timestamps
+with notes. These will not include the fixed "First Published" change note as
+this will automatically appended to the start of a change history with a time
+based on the `first_published_at` value for a document (which can be overriden
+by a contents backdating) and therefore does not have a fixed time. It also
+does not include the current change note of an edition because this not become
+an aspect of change history until the content is published and the update type
+and publishing time are fixed. It was noted that having both `change_note`
+and `change_history` fields risks confusion for developers, however this
+seemed preferable to complicating the `change_history` field with additional
+responsibilities.
+
+Initially the only means to edit items in a change history will be via a rake
+task that developers can perform. This will provide means to add, edit and
+delete items. Eventually we may choose to build the means to edit this into the
+Web UI however at this point in time it is unclear which users would be granted
+the permissions to do this given the risks this editing can have on the
+integrity of GOV.UK content.
+
+We will expand the [IntegrityChecker][] system used by Content Publisher to
+verify that times and change history sent to the Publishing API matches what
+is already stored for Whitehall content. This will be used to ensure that
+we have a confidence these will remain consistent through a migration.
+
+We will remove the deprecation of the `change_history` attribute in Publishing
+API. This is to reflect that this field is still in operation and that the
+recommended alternative approach, of using the `change_note` field, does not
```

Comment:
This is a wider question but is something that occurred to me when reading this - If we're removing the deprecation from change_history, should we be adding one to the change_note field? It feels like if we've decided that change_history offers a lot more then this should be the preferred approach to managing change notes.

Reply:
Okay sure that makes sense, feels like we should maybe add something somewhere in the publishing-api to advise future devs looking into this topic that change_history is the field that should ideally be used.

My original thought process was that a deprecation warning on the change_note field would give that impression but maybe there is some other way to convey that intent, what do you think?

Reply:
I don't think it's necessarily ideal though. If you're doing basic publishing the change_note field is simpler to use and a good choice. If you want to support advanced features like backdating or removing a change note you have to use change_history with the added burden of date management.